pkg/descheduler/strategies/topologyspreadconstraint.go

```diff
@@ -0,0 +1,242 @@
+				// If the diff is within the skew, we move to the next "belowAvg" topology.
+				// If the "aboveAvg" topology can't give any more without falling below the average, we move to the next "aboveAvg"
+				//
+				// Following this, the above topologies end up "sorted" as:
+				//  5  5
+				//  5  5
+				//  5
+				//  5
+				// (assuming even distribution by the scheduler of the evicted pods)
+				idealAvg := int(math.Ceil(sumPods / float64(len(constraintTopologies))))
+				belowAvg, aboveAvg := splitTopologiesAroundAverage(constraintTopologies, idealAvg)
+
+				// i is the index for belowAvg
+				// j is the index for aboveAvg
+				i, j := 0, 0
+				for i < len(belowAvg) && j < len(aboveAvg) {
+					// if j has no more to give without falling below the ideal average, move to next aboveAvg
+					if len(aboveAvg[j].pods)-idealAvg <= 0 {
+						j++
+					}
+
+					// skew = diff - maxSkew
+					skew := math.Abs(float64(len(belowAvg[i].pods)-len(aboveAvg[j].pods))) - float64(constraint.MaxSkew)
+
+					// if i and j are within the maxSkew of each other, move to next belowAvg
+					if int32(skew) <= constraint.MaxSkew {
+						i++
+						continue
+					}
+
+					// the most that can be given from aboveAvg is up to halfway,
+					// or how many it has remaining without falling below the average
+					movePods := int(math.Min(math.Ceil(skew/2), float64(len(aboveAvg[j].pods)-idealAvg)))
+					if movePods == 0 {
+						i++
+						continue
+					}
+
+					// remove pods from aboveAvg and add them to the list of pods to be evicted
+					// also (just for tracking), add them to the list of pods in belowAvg[i]
+					aboveToEvict := aboveAvg[j].pods[movePods-1:]
+					for _, pod := range aboveToEvict {
+						podsForEviction[pod] = struct{}{}
+					}
+					aboveAvg[j].pods = aboveAvg[j].pods[:movePods]
+					belowAvg[i].pods = append(belowAvg[i].pods, aboveToEvict...)
+				}
+			}
+		}
+	}
+
+	for pod := range podsForEviction {
+		node, err := client.CoreV1().Nodes().Get(ctx, pod.Spec.NodeName, metav1.GetOptions{})
+		if err != nil {
+			klog.ErrorS(err, "couldn't get node", "node", pod.Spec.NodeName)
+		}
+		if _, err := podEvictor.EvictPod(ctx, pod, node, "PodTopologySpread"); err != nil {
+			klog.ErrorS(err, "Error evicting pod", "pod", klog.KObj(pod))
+			break
+		}
+	}
+}
+
+// splitTopologiesAroundAverage sorts and splits the list of topologies based on their size
+// It returns 2 lists: topologies <= average (ascending) and topologies > average (descending)
+func splitTopologiesAroundAverage(constraintTopologies map[topologyPair][]*v1.Pod, idealAvg int) ([]topology, []topology) {
+	sortedTopologies := make([]topology, 0, len(constraintTopologies))
+	// sort the topologies and return 2 lists: those <= the average and those > the average (> list inverted)
+	for pair, list := range constraintTopologies {
+		sortedTopologies = append(sortedTopologies, topology{pair: pair, pods: list})
+	}
+
+	// create an ascending slice of all topologies
```

Comment:
"... all topology values". Topology keys is constant here as `splitTopologiesAroundAverage` is called within a single constraint. Each constraint has only a single topology key.

Reply:
Also, worth renaming `constraintTopologies` to `constraintTopologyKeyValues` or similar to stress the topology key is constant.

Reply:
Yeah, I wasn't sure on the terminology so I probably used "topology" to ambiguously refer to either an entire topology (all its pairs) and a single topology zone